File: source/glparts/Texture.java
package glparts;

import callback.EmptyCallback;
import glparts.abs.GLObject;
import math.Int2;
import named.NameHolder;
import utility.Instance;
import utility.Memory;
import utility.helper.FileHelper;
import window.GLContext;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.Serial;
import java.io.Serializable;
import java.nio.ByteBuffer;
import java.util.Arrays;

import static org.lwjgl.opengl.GL33.*;

public class Texture extends GLObject implements Serializable {
    private final Int2 size;
    private final byte[] pixels;
    private transient int buffer;

    public Texture(NameHolder holder, String name, GLContext context, Int2 size, byte[] pixels) {
        super(holder, name, context);

        this.size = new Int2(size);
        this.pixels = Instance.isValid(pixels) ? Arrays.copyOf(pixels, pixels.length) : null;

        buffer = generateTexture(size, pixels);
    }

    public Texture(NameHolder holder, String name, GLContext context, String filepath, boolean flipY) {
        this(holder, name, context, FileHelper.getImageSize(filepath), FileHelper.getImageData(filepath, flipY));
    }

    public Texture(NameHolder holder, String name, GLContext context, String filepath) {
        this(holder, name, context, filepath, true);
    }

    @Override
    public void deallocate() {
        glDeleteTextures(buffer);
        buffer = 0;
    }

    @Override
    public boolean isAllocated() {
        return buffer != 0;
    }

    @Serial
    private void readObject(ObjectInputStream stream) throws IOException, ClassNotFoundException {
        stream.defaultReadObject();
        buffer = generateTexture(size, pixels);
    }

    public void setWrap(int wrapS, int wrapT) {
        glBindTexture(GL_TEXTURE_2D, buffer);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, wrapS);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, wrapT);
        glBindTexture(GL_TEXTURE_2D, 0);
    }

    public void setFilters(int min, int mag) {
        glBindTexture(GL_TEXTURE_2D, buffer);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, min);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, mag);
        glBindTexture(GL_TEXTURE_2D, 0);
    }

    public void use(int slot, EmptyCallback callback) {
        glActiveTexture(GL_TEXTURE0 + slot);
        glBindTexture(GL_TEXTURE_2D, buffer);

        callback.method();

        glActiveTexture(GL_TEXTURE0 + slot);
        glBindTexture(GL_TEXTURE_2D, 0);

        glActiveTexture(GL_TEXTURE0);
    }

    public int getBuffer() {
        return buffer;
    }

    public Int2 getSize() {
        return new Int2(size);
    }

    public void resize(Int2 size) {
        if (!this.size.equals(size)) {
            glBindTexture(GL_TEXTURE_2D, buffer);
            glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA32F, size.x, size.y, 0, GL_RGBA, GL_UNSIGNED_BYTE, (ByteBuffer) null);
            glBindTexture(GL_TEXTURE_2D, 0);
            this.size.set(size);
        }
    }

    public static int generateTexture(Int2 size, byte[] pixels) {
        int buffer = glGenTextures();

        glBindTexture(GL_TEXTURE_2D, buffer);

        glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA32F, size.x, size.y, 0, GL_RGBA, GL_UNSIGNED_BYTE, Memory.createByteBuffer(pixels));

        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);

        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);

        glBindTexture(GL_TEXTURE_2D, 0);

        return buffer;
    }

    public static void use(int[] textures, EmptyCallback callback) {
        for (int i = 0; i < textures.length; i++) {
            glActiveTexture(GL_TEXTURE0 + i);
            glBindTexture(GL_TEXTURE_2D, textures[i]);
        }

        callback.method();

        for (int i = 0; i < textures.length; i++) {
            glActiveTexture(GL_TEXTURE0 + i);
            glBindTexture(GL_TEXTURE_2D, 0);
        }

        glActiveTexture(GL_TEXTURE0);
    }
}

File: source/glparts/Texture.kt
package glparts

import glparts.abs.GLObject
import math.Int2
import named.NameHolder
import org.lwjgl.opengl.GL11
import org.lwjgl.opengl.GL13
import org.lwjgl.opengl.GL30
import utility.Memory
import utility.helper.FileHelper
import window.GLContext
import java.io.IOException
import java.io.ObjectInputStream
import java.io.Serial
import java.io.Serializable
import java.nio.ByteBuffer

class Texture(holder: NameHolder?, name: String?, context: GLContext?, size: Int2, pixels: ByteArray?) : GLObject(holder, name, context), Serializable {
    private val size: Int2
    private val pixels: ByteArray?

    @Transient
    var buffer: Int
        private set

    init {
        this.size = Int2(size)
        this.pixels = pixels?.copyOf()
        buffer = generateTexture(size, pixels)
    }

    constructor(holder: NameHolder?, name: String?, context: GLContext?, filepath: String, flipY: Boolean = true) : this(holder, name, context, FileHelper.getImageSize(filepath), FileHelper.getImageData(filepath, flipY)!!)

    override fun deallocate() {
        GL11.glDeleteTextures(buffer)
        buffer = 0
    }

    override fun isAllocated(): Boolean {
        return buffer != 0
    }

    @Serial
    @Throws(IOException::class, ClassNotFoundException::class)
    private fun readObject(stream: ObjectInputStream) {
        stream.defaultReadObject()
        buffer = generateTexture(size, pixels)
    }

    fun setWrap(wrapS: Int, wrapT: Int) {
        GL11.glBindTexture(GL11.GL_TEXTURE_2D, buffer)
        GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_WRAP_S, wrapS)
        GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_WRAP_T, wrapT)
        GL11.glBindTexture(GL11.GL_TEXTURE_2D, 0)
    }

    fun setFilters(min: Int, mag: Int) {
        GL11.glBindTexture(GL11.GL_TEXTURE_2D, buffer)
        GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_MIN_FILTER, min)
        GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_MAG_FILTER, mag)
        GL11.glBindTexture(GL11.GL_TEXTURE_2D, 0)
    }

    fun use(slot: Int, callback: () -> Unit) {
        GL13.glActiveTexture(GL13.GL_TEXTURE0 + slot)
        GL11.glBindTexture(GL11.GL_TEXTURE_2D, buffer)
        callback()
        GL13.glActiveTexture(GL13.GL_TEXTURE0 + slot)
        GL11.glBindTexture(GL11.GL_TEXTURE_2D, 0)
        GL13.glActiveTexture(GL13.GL_TEXTURE0)
    }

    fun getSize(): Int2 {
        return Int2(size)
    }

    fun resize(size: Int2) {
        if (!this.size.equals(size)) {
            GL11.glBindTexture(GL11.GL_TEXTURE_2D, buffer)
            GL11.glTexImage2D(GL11.GL_TEXTURE_2D, 0, GL30.GL_RGBA32F, size.x, size.y, 0, GL11.GL_RGBA, GL11.GL_UNSIGNED_BYTE, null as ByteBuffer?)
            GL11.glBindTexture(GL11.GL_TEXTURE_2D, 0)
            this.size.set(size)
        }
    }

    companion object {
        fun generateTexture(size: Int2, pixels: ByteArray?): Int {
            val buffer = GL11.glGenTextures()
            GL11.glBindTexture(GL11.GL_TEXTURE_2D, buffer)
            GL11.glTexImage2D(GL11.GL_TEXTURE_2D, 0, GL30.GL_RGBA32F, size.x, size.y, 0, GL11.GL_RGBA, GL11.GL_UNSIGNED_BYTE, if (pixels != null) Memory.createByteBuffer(pixels) else null)
            GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_MIN_FILTER, GL11.GL_LINEAR)
            GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_MAG_FILTER, GL11.GL_LINEAR)
            GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_WRAP_S, GL11.GL_REPEAT)
            GL11.glTexParameteri(GL11.GL_TEXTURE_2D, GL11.GL_TEXTURE_WRAP_T, GL11.GL_REPEAT)
            GL11.glBindTexture(GL11.GL_TEXTURE_2D, 0)
            return buffer
        }

        fun use(textures: IntArray, callback: () -> Unit) {
            for (i in textures.indices) {
                GL13.glActiveTexture(GL13.GL_TEXTURE0 + i)
                GL11.glBindTexture(GL11.GL_TEXTURE_2D, textures[i])
            }
            callback()
            for (i in textures.indices) {
                GL13.glActiveTexture(GL13.GL_TEXTURE0 + i)
                GL11.glBindTexture(GL11.GL_TEXTURE_2D, 0)
            }
            GL13.glActiveTexture(GL13.GL_TEXTURE0)
        }
    }
}